### ISSUE TITLE
BusyBox build failed on Ubuntu 16.04 64-bit default installation

### ISSUE THREAD

**Issue by slavaim** [ivandavidov/minimal#54]
BusyBox build failed on Ubuntu 16.04 64-bit default installation. The build failed on two Ubuntu 16.04 systems. Ubuntu 14 was not affected by this issue. 

The error is below

```
In file included from /usr/lib/gcc/x86_64-linux-gnu/5/include-fixed/syslimits.h:7:0,
                 from /usr/lib/gcc/x86_64-linux-gnu/5/include-fixed/limits.h:34,
                 from include/platform.h:141,
                 from include/libbb.h:13,
                 from include/busybox.h:8,
                 from applets/applets.c:9:
/usr/lib/gcc/x86_64-linux-gnu/5/include-fixed/limits.h:168:61: error: no include path in which to search for limits.h
In file included from include/libbb.h:13:0,
                 from include/busybox.h:8,
                 from applets/applets.c:9:
include/platform.h:152:23: fatal error: byteswap.h: No such file or directory
compilation terminated.
scripts/Makefile.build:197: recipe for target 'applets/applets.o' failed
make[1]: *** [applets/applets.o] Error 1
Makefile:372: recipe for target 'applets_dir' failed
make: *** [applets_dir] Error 2
Generating BusyBox based initramfs area...
  CC      applets/applets.o
In file included from /usr/lib/gcc/x86_64-linux-gnu/5/include-fixed/syslimits.h:7:0,
                 from /usr/lib/gcc/x86_64-linux-gnu/5/include-fixed/limits.h:34,
                 from include/platform.h:141,
                 from include/libbb.h:13,
                 from include/busybox.h:8,
                 from applets/applets.c:9:
/usr/lib/gcc/x86_64-linux-gnu/5/include-fixed/limits.h:168:61: error: no include path in which to search for limits.h
In file included from include/libbb.h:13:0,
                 from include/busybox.h:8,
                 from applets/applets.c:9:
include/platform.h:152:23: fatal error: byteswap.h: No such file or directory
```

**Comment by AwlsomeAlex** [comment]
It could be something to do with the Level of Optimization, due to the current option optimizing for size, maybe corrupting this. Try changing line 64 of "07_build_busybox" where it says `# compile busybox' from `EXTRA_CFLAGS="-Os -s -fno-stack-protector -U_FORTIFY_SOURCE" \` to `  EXTRA_CFLAGS="-O2 -s -fno-stack-protector -U_FORTIFY_SOURCE" \` stated [here](https://wiki.gentoo.org/wiki/GCC_optimization#-O). Your code might be a bit bigger, but better than not having it working at all. Hope this helps.

**Comment by ivandavidov** [comment]
I have tried MLL on Ubuntu 16.04 and it works fine. Nevertheless, you can try to remove the ``EXTRA_CFLAGS`` line and then try to build the ISO image again. Even though we are using the same Ubuntu OS, there might be differences in the GCC versions that we use, as well as differences in other build chain tools.

One more thing - try to do ``sudo apt update`` + ``sudo apt upgrade`` + ``sudo apt dist-upgrade`` in order to ensure that you are using the latest software packages on your Ubuntu OS. If it fails again, then you can try to remove the ``EXTRA_CFLAGS`` line.

**Comment by slavaim** [comment]
Did you build on 32-bit or 64-bit system?
I tried the above solutions with system update and compilers flags. None has helped. The compilation failed. 
I even tried to switch the GCC version back to 4.8 as on Ubuntu 14.

**Comment by ivandavidov** [comment]
My physical machine has 64-bit Ubuntu and back in April everything was fine. However, since then I've been playing mostly with Linux Mint (both 32-bit and 64-bit).

For now I suggest to disable the mentioned compiler flags completely. It's the safest way to ensure that the issue is not related to GCC and the build chain.

**Comment by slavaim** [comment]
Disabling the additional compiler flags didn't help also. I believe the problem with the build environment on Ubuntu 16.04 64-bit. Googling it shows that it is not unique and a tricky one. My main system is Ubuntu 14 64-bit which is safe from this problem, but I haven't had any problem on Ubuntu 16.04 64-bit except BusyBox compilation for ML. I was able to build kernels and QEMU on these Ubuntu 16.04 64-bit systems, so I assume these systems were in a sane state.

**Comment by ivandavidov** [comment]
Did you check the generated output from ``02_build_kernel.sh`` and ``05_prepare_glibc.sh``? Do you have the kernel headers generated? And do you have the new symbolic links generated in the GLIBC space?

Assuming everything is OK - can you build BusyBox manually by invoking ``make menuconfig``, then ``make busybox`` and finally ``make install``? If you can manually build it, then the issue is not in GCC itself but somewhere in the kernel header linking as I suggested before.

**Comment by slavaim** [comment]
Yes, I was able to build a local standalone build with 
```
make defconfig
make
```

Running `02_build_kernel.sh` completed successfully and as far as I can tell the headers were generated correctly. I am not sure about  `05_prepare_glibc.sh` but no errors were reported.

**Comment by ivandavidov** [comment]
OK, last but not least - does it fail if you change the kernel and/or busybox version? For example, try to switch to the latest long-term kernel release (4.4.40 at this point) and the latest stable busybox (1.26.1 at this point). You might want to give it a try with several different older and newer kenel/busybox combinations.

**Comment by ivandavidov** [comment]
@slavaim - I just tested the build process with new installation of the latest available Ubuntu distribution with all applied updates. Then I installed the dependencies described on the [official MLL website](http://minimal.idzona.com). I also changed the main ``.config`` file in MLL and updated the kernel to long-term version 4.4.40, and BusyBox to version 1.26.1.

There were no issues at all. I have to admit that due to lack of enough time I only checked this use case and I didn't check what will happen if I use the latest stable kernel instead of the latest long-term kernel.

Since the long-term kernel is what 99% of all Linux distributions use, from now on in my official sources and builds I will also stick with the latest available long-term kernel sources. I have already pushed the the MLL configuration changes in the repository.

If you still experience the issue, even though you are using the same kernel/BB versions which I used (kernel 4.4.40 and BusyBox 1.26.1), then the issue must be local. In this case it is up to you to figure out what's wrong. I'm sorry that I can't assist you more on this matter.

Good luck!
